java/ql/lib/semmle/code/java/dataflow/ListOfConstantsSanitizer.qll

```diff
@@ -0,0 +1,357 @@
+class CollectionClass extends string {
+  CollectionClass() { this = ["List", "Set"] }
+}
+
+/** Classes for `java.util.List` and `java.util.Set`. */
+module Collection {
+  private class CollectionContainsCall extends MethodCall {
+    CollectionClass collectionClass;
+
+    /** Gets whether the collection is a "List" or a "Set". */
+    CollectionClass getCollectionClass() { result = collectionClass }
+
+    CollectionContainsCall() {
+      exists(Method m |
+        this.getMethod() = m and
+        m.hasName("contains") and
+        m.getDeclaringType()
+            .getSourceDeclaration()
+            .getASourceSupertype*()
+            .hasQualifiedName("java.util", collectionClass)
+      )
+    }
+  }
+
+  /**
+   * A call with a collection of class `collectionClass` as argument `arg` (or
+   * as qualifier, if `arg` is -1) which will not add any new non-constant
+   * elements to it.
+   */
+  abstract class SafeCall extends Call {
+    int arg;
+    CollectionClass collectionClass;
+
+    SafeCall() {
+      arg = -1 and exists(this.getQualifier())
+      or
+      exists(this.getArgument(arg))
+    }
+
+    /** Gets whether the collection is a "List" or a "Set". */
+    CollectionClass getCollectionClass() { result = collectionClass }
+
+    /** Gets the argument index, or -1 for the qualifier. */
+    int getArg() { result = arg }
+  }
+
+  private class AddConstantElement extends SafeCall, MethodCall {
+    AddConstantElement() {
+      arg = -1 and
+      exists(Method m, RefType t |
+        this.getMethod() = m and
+        t = m.getDeclaringType().getSourceDeclaration().getASourceSupertype*()
+      |
+        collectionClass = "List" and
+        t.hasQualifiedName("java.util", "List") and
+        m.getName() = ["add", "addFirst", "addLast"] and
+        this.getArgument(m.getNumberOfParameters() - 1).isCompileTimeConstant()
+        or
+        collectionClass = "Set" and
+        t.hasQualifiedName("java.util", "Set") and
+        m.getName() = "add" and
+        this.getArgument(0).isCompileTimeConstant()
+      )
+    }
+  }
+
+  private class UnmodifiableCollection extends SafeCall, MethodCall {
+    UnmodifiableCollection() {
+      arg = 0 and
+      exists(Method m |
+        this.getMethod() = m and
+        m.hasName("unmodifiable" + collectionClass) and
+        m.getDeclaringType()
+            .getSourceDeclaration()
+            .getASourceSupertype*()
+            .hasQualifiedName("java.util", "Collections")
+      )
+    }
+  }
+
+  /**
+   * Gets an `Expr` which locally flows to `e` and which nothing locally flows
+   * to.
+   */
+  private Expr getALocalExprFlowRoot(Expr e) {
+    DataFlow::localExprFlow(result, e) and
+    not exists(Expr e2 | e2 != result | DataFlow::localExprFlow(e2, result))
+  }
+
+  /**
+   * Holds if `e` was not involved in any calls which might add non-constant
+   * elements.
+   */
+  private predicate noUnsafeCalls(Expr e) {
```

Comment:
This predicate at a glance looks dubious from a performance perspective. Also, the "not involved" part of the qldoc indicates that the semantics of this predicate is somewhat unclear. And the entire thing - both the name and the predicate body is a top-level negation, which indicates that it might be better to represent the opposite.